### ISSUE TITLE
Public methods whose parameters and return types are not public

### ISSUE THREAD

**Issue by lingocoder** [bitcoinj/bitcoinj#1873]
## Context

I'm developing [*an ABI inspection tool*](https://bit.ly/AbiImpl). I call it *The ABI*. It scans the project classes and dependencies of Java libraries. From the scan, it reports on classes whose public API exposes types from the library's dependencies.

I did a scan of the bitcoinj-core library. The abi reported several classes
 that it determined exports types from different dependencies through the library's public API. ([*attached*](https://github.com/bitcoinj/bitcoinj/files/3238989/bitcoinj.abi.report.0.b.txt))

The ABI inspector's results jibe with [*Gradle's guidelines for separating `api` and `implementation` dependencies*](http://bit.ly/ImplApiSep) into appropriate configurations. However the results differ from the transitivity currently defined for bitcoinj-core's dependencies.

The reason I'm engaging is because I'm looking to compare the results arrived at by the tool with bitcoinj authors' actual expectations for the transitivity of the library's dependencies. I'm hoping your feedback will help improve the accuracy of the abi's results.

The following table lists what the configurations are currently and what the inspection suggests that they could be changed to:

|      |   Currently    |   Analyzed as   |
| :--- | :--------------: | :-------------: |
| *`org.bouncycastle:bcprov-jdk15on:1.60`* | compile | api |
| *`com.squareup.okhttp3:okhttp:3.12.3`* | implementation | api |
| *`com.google.protobuf:protobuf-java:3.7.1`* | compile | api |
| *`com.google.guava:guava:27.1-android`* | implementation | api |
| *`net.jcip:jcip-annotations:1.0`* | implementation | api |  

The analysis the tool applies is [*based on Gradles recommended best practices*](http://bit.ly/iDimpLApi) for assigning appropriate configurations to dependencies.

## How the ABI Inspector's analysis was verified

1. Download bitcoinj-master.zip from Code page, unzip and cd to install dir
2. `gradle -x javadoc clean assemble`
3. Install the jar that is built to your local Maven repository (*assumes Maven is installed*)

        mvn org.apache.maven.plugins:maven-install-plugin:3.0.0-M1:install-file \
        -Dfile=${install_dir}\bitcoinj-master\core\build\libs\bitcoinj-core-0.16-SNAPSHOT.jar \
        -DgroupId=org.bitcoinj -DartifactId=bitcoin-core \
        -Dversion=0.16-SNAPSHOT -Dpackaging=jar \
        -DlocalRepositoryPath=${path_to_maven_local_repo}

4. Create a new, standalone Gradle application project in a different location
5. Add the locally-published bitcoinj artifact to the *`dependencies{}`* block of the new project's *`build.gradle`* with *`org.bitcoinj:bitcoinj-core:0.16-SNAPSHOT`*
6. Write code in the new project that uses the following classes:

    - *`org.bouncycastle.crypto.params.KeyParameter`* (*from org.bouncycastle:bcprov-jdk15on:1.60*)
    - ~~*`okhttp3.OkHttpClient`* (*from com.squareup.okhttp3:okhttp:3.12.3*)~~
    - *`com.google.protobuf.ByteString`* (*from com.google.protobuf:protobuf-java:3.7.1*)
    - ~~*`com.google.common.util.concurrent.ListenableFuture`* (*from com.google.guava:guava:27.1-android*)~~

7. Compile it. ([*example*](http://bit.ly/btcAbiPoc))

## Expected behavior

If a method of a library is public, I expect its parameters and return types to be public too. If dependencies are intended to be available to consumers, I would expect that they'd be explicitly included in Gradle's *`api`* configuration.

Mingling of the deprecated *`compile`* configuration right along-side the *`implementation`* configuration that is meant to supersede it, is an unexpected inconsistency for a library of bitcoinj-core's quality. I expected conformance to best practice.

## Actual Behavior

The types that are lined through in the bullet points above are interesting. They're in the *`implementation`* configuration. However, they are present in the library's public API; either as public method arguments, return types or public constructor arguments. So a consumer has access to the public methods of the library's API, but they don't have access to those public method's inputs and outputs. Public methods whose signatures are hidden? Is that intentional?

[***The attached project demonstrates***](http://bit.ly/btcAbiPoc) apparent anomalies like *`okhttp3.OkHttpClient`* and *`ListenableFuture`* that are either returned from or passed into public API methods or constructors, but are hidden from consumers. They are commented out. But you'll get the gist.

## TL;DR

My basic question is: Is the transitivity expressed in bitcoinj's Gradle build script, intentionally different from the actual transitivity of the dependencies exposed by the library's public API? Please see ([*the report*](https://github.com/bitcoinj/bitcoinj/files/3238989/bitcoinj.abi.report.0.b.txt)) +  ([*the example*](http://bit.ly/btcAbiPoc)) for context.

**Comment by msgilligan** [comment]
> My basic question is: Is the transitivity expressed in bitcoinj's Gradle build script, intentionally different from the actual transitivity of the dependencies exposed by the library's public API? 

I'm pretty sure the transitivity expressed in the Gradle script is unintentional.

Your tool looks useful and I think it can help us improve **bitcoinj**. I've been personally trying to remove unnecessary dependencies in both the implementation and the API (and I'm also starting to work on using Java Platform Module System) so I'm broadly supportive of work that will reduce our external dependencies and more clearly express our intentions with respect to those dependencies.

The trade-offs are made more difficult because the project supports both Android and JavaSE, so we can't use Java 8 (or later) types and have strong historical dependencies on Guava.

I'm looking forward to hearing @schildbach 's comments on this issue.

**Comment by lingocoder** [comment]
> _...It would be nice if the report included summary information (number of API methods that use each library) so we could track our progress over time..._.

Hey there.  Thanks for the feedback :+1:

That is a good idea to be able to do that. The inspection process does use that info already. I will think about the most efficient way to pull it into the report.

> _...we can't use Java 8 (or later) types and have strong historical dependencies on Guava..._

Got it. Thanks. 

I should point out that the abi inspection is agnostic as far as what the specific types are that it analyzes. Likewise, the class file version of the types it analyzes never comes into the picture. That is, not unless I run into some gotcha down the road that I haven't up until now. You've given me some unit test cases I should consider adding. Thanks again.

----

Can you help me out? As a consumer of the bitcoinj-core library, should I be able to do something like this?

        NetworkParameters regTest = NetworkParameters
                .fromID( NetworkParameters.ID_REGTEST );

        Wallet testWallet = Wallet.createBasic( regTest );

        Coin coin = testWallet.getBalance( );

        Details details = new Details( ... );

        /** ABI Dependency : okhttp */
        OkHttpClient okHttpClient = new OkHttpClient( ); 
        HttpDiscovery httpDiscovery = new HttpDiscovery( regTest, details, okHttpClient ); 
        ....
        /** ABI Dependency : bcprov-jdk15on */
        KeyParameter userKey = new KeyParameter( ... );
        ...
        PaymentChannelClient pcClient = new PaymentChannelClient( testWallet...);
        ...
        /** ABI Dependency : protobuf-java */
        ByteString info = ByteString.EMPTY;
        ...
        /** ABI Dependency : guava */
        ListenableFuture<?> guavaFuture = pcClient.incrementPayment( coin, info, userKey ); 
        ...

I admit it is a naively contrived case. But, just from an API point of view? Is there anything in the above example that is never intended to be written by bitcoinj-core library-consuming applications?

**Comment by schildbach** [comment]
I think @msgilligan summed it up quite nicely.

From your example, I'd say I'd like to hide OkHttp, BouncyCastle and Protobuf from the consumer, but at the same time I'm aware that this currently isn't the case. For example, for the recent SpongyCastle to BouncyCastle migration I had to change quite a number of imports for my bitcoinj-consuming app.

Guava I'd say is part of the API and that feels ok to me.

**Comment by msgilligan** [comment]
> Guava I'd say is part of the API and that feels ok to me.

I think we should look at the use go Guava types in the API on a case-by-case basis. There are certainly some places (e.g. `ImmutableList<ChildNumber>`) where we could easily migrate away from Guava (e.g. to `List` or even a custom type say `HDPath`) and others (e.g. `ListenableFuture`) where there is no JDK 7 type that is equivalent. When we move to a minimum JDK 8 we could migrate to `CompletableFuture`, etc.

On a separate subject using the Java Platform Module System will allow us to stop exporting internal classes that have `public` methods with dependent types that we do not want to make part of the API. At some point in the future I'd like to investigate making the `core` JAR a multi-release JAR.

**Comment by lingocoder** [comment]
Hey there.

In [*the latest iteration of abi.cli*](http://bit.ly/AbiImpl), I have now implemented an option that shows what specific types are used. That's not precisely the method count summary that you suggested. I know. But it does now show hopefully equally-useful info on what specifically is being used from specific dependencies. 

I'm now showing stuff like specific super types, method/constructor names, parameters, exceptions and return types.

Here are some examples ([*more attached*](https://github.com/bitcoinj/bitcoinj/files/3268236/bitcoinj.issue.1873.abi.cli.execjar.0.log)):

````
>--: ABI dependencies found for...
|
|__ class : org.bitcoinj.protocols.channels.PaymentChannelV2ServerState$1
|   |__ supertype : com.google.common.util.concurrent.FutureCallback
|
|
|_________ Dependencies ________
|
|* guava-27.1
|_______________________________

...
>--: ABI dependencies found for...
|
|__ class : org.bitcoinj.core.Peer$GetDataRequest
|   |__ constructor : org.bitcoinj.core.Peer$GetDataRequest
|     |__ param : com.google.common.util.concurrent.SettableFuture
|
|
|_________ Dependencies ________
|
|* guava-27.1
|_______________________________

...
>--: ABI dependencies found for...
|
|__ class : org.bitcoinj.net.discovery.HttpDiscovery
|   |__ constructor : org.bitcoinj.net.discovery.HttpDiscovery
|     |__ param : okhttp3.OkHttpClient
|   |__ method : protoToAddrs
|     |__ exception : com.google.protobuf.InvalidProtocolBufferException
|
|
|_________ Dependencies ________
|
|* okhttp
|* protobuf-java
|_______________________________

...
>--: ABI dependencies found for...
|
|__ class : org.bitcoinj.protocols.channels.PaymentChannelClientConnection
|   |__ constructor : org.bitcoinj.protocols.channels.PaymentChannelClientConnection
|     |__ param : org.bouncycastle.crypto.params.KeyParameter
|   |__ method : incrementPayment
|     |__ return : com.google.common.util.concurrent.ListenableFuture
|   |__ method : incrementPayment
|     |__ param : com.google.protobuf.ByteString
|     |__ param : org.bouncycastle.crypto.params.KeyParameter
|     |__ return : com.google.common.util.concurrent.ListenableFuture
|   |__ method : getChannelOpenFuture
|     |__ return : com.google.common.util.concurrent.ListenableFuture
|
|
|_________ Dependencies ________
|
|* bcprov-jdk15on
|* guava-27.1
|* protobuf-java
|_______________________________

...

>--: ABI dependencies found for...
|
|__ class : org.bitcoinj.crypto.DeterministicHierarchy
|   |__ method : deriveNextChild
|     |__ param : com.google.common.collect.ImmutableList
|   |__ method : getNumChildren
|     |__ param : com.google.common.collect.ImmutableList
|
|
|_________ Dependencies ________
|
|* guava-27.1
|_______________________________

...

````
I would like the display to be prettier. That's on my *TODO* list.

Any other suggestions or questions are welcomed. Thanks.

> „*...At some point in the future I'd like to investigate making the core JAR a multi-release JAR...*“

[*I recently made a Gradle plugin of mine a multi-release JAR*](http://bit.ly/mrJrPlug). I would recommend doing it. For sure. There is a certain amount of additional maintenance effort involved for each different supported version though. That shouldn't be underestimated. But as long as it's planned well and given enough forethought, I think every library with significant maturity and stability should do it. 

Less-mature libraries, on the other hand, that are in more of a state of flux might want to hold fire until their thing stabilizes.

**Comment by msgilligan** [comment]
PR #1877 removes `com.google.common.collect.ImmutableList` from the API for a couple of public classes/methods.

**Comment by msgilligan** [comment]
See related Issue #1963.

**Comment by msgilligan** [comment]
@lingocoder Would you be up for making a branch of bitcoinj that runs your analyzer from a Gradle task? I don't know if we'd want to merge it into `master`, but we could rebase it periodically and run it to track our api dependencies.

**Comment by lingocoder** [comment]
Hi @msgilligan,

Apologies for the delayed reply. Sure you bet. I might have some time in the next couple days to do that.

I'll let you know when it's there.

**Comment by msgilligan** [comment]
Note that PR #1965 was merged making Java an `api` dependency.

**Comment by msgilligan** [comment]
PR #2198 will make OkHttp an `implementation` not an `api` dependency (technically after one `@Deprecated` constructor is removed)
